### ISSUE TITLE
CWG2687 [over.match.call.general] Calling explicit object member functions via addresses of overload sets

### ISSUE THREAD

**Issue by LegionMammal978** [cplusplus/CWG#225]
Full name of submitter (unless configured in github; will be published with the issue): Matthew House

Reference (section label): [[over.match.call.general] paragraph 2](https://wg21.link/over.match.call.general#2)

Link to reflector thread (if any):

Issue description:

According to [[over.match.call.general] paragraph 2](https://wg21.link/over.match.call.general#2),

> If the [*postfix-expression*](https://wg21.link/expr.post.general#nt:postfix-expression) is the address of an overload set, overload resolution is applied using that set as described above. If the function selected by overload resolution is a non-static member function, the program is ill-formed.

This wording implies that an address expression referring to an explicit object member function may not be called directly, but only via an intermediate pointer value:

```c++
struct A { void b(this A) {} };
int main() {
  static_cast<void (*)(A)>(&A::b)(A()); // valid
  (&A::b)(A());                         // invalid
}
```

There seems to be no particular reason for why this should be the case, given that the function pointer obtained from the address expression can only be called with the object parameter in the argument list.

Suggested resolution:

Change [[over.match.call.general] paragraph 2](https://wg21.link/over.match.call.general#2) as follows:

> If the [*postfix-expression*](https://wg21.link/expr.post.general#nt:postfix-expression) is the address of an overload set, overload resolution is applied using that set as described above. If the function selected by overload resolution is a<ins>n implicit object</ins> <del>non-static</del> member function, the program is ill-formed.

**Comment by jensmaurer** [comment]
CWG2687

**Comment by cpplearner** [comment]
Consider

```c++
struct A {
    static void f(A);
    void f(this A);

    void g();
};

void A::g() {
    (&A::f)(A()); // #1
    (&A::f)();    // #2
}
```

Currently `#1` calls the static member function and `#2` is ill-formed. Do you intend to make `#2` well-formed? What if `A::g` is a non-member function?

**Comment by LegionMammal978** [comment]
I have no intent to make `#2` well-formed: the function pointer can never take an implied argument. But I hadn't realized that `#1` would be well-formed. There might be some confusion here, since `static void f(A)` has an additional implicit object parameter, but `void f(this A)` does not. How does overload resolution select between the two candidates when they have argument lists of different length? Does [[over.call.func] para. 3](https://wg21.link/over.call.func#3) apply when calling the address of an overload set?

What I find even more odd about your example is that MSVC successfully resolves `&A::f` to `void f(this A)` in non-call contexts such as `void (*fp)(A) = &A::f;`. By my reading of [[over.over]](https://wg21.link/over.over), both functions are selected and the expression should be ambiguous.

---

Also, I think there might be a typo in [[over.over] para. 4](https://wg21.link/over.over#4):

> Non-member functions, static member functions, and explicit object member functions match targets of function pointer type or reference to function type. Non-static member functions match targets of pointer-to-member-function type.

"Non-static member functions" should read "implicit object member functions" per [P0847](https://wg21.link/p0847), unless explicit object member functions are also supposed to match pointer-to-member-function targets.

**Comment by jicama** [comment]
I think the declarations of the two functions should be ill-formed as an invalid overload.  We could do that by changing "non-object-parameter-type-list" to "parameter-type-list or non-object-parameter-type-list" in [basic.scope.scope].

**Comment by jensmaurer** [comment]
> Also, I think there might be a typo in [[over.over] para. 4](https://wg21.link/over.over#4):

Yes, that's a missed edit when applying the original paper. See cplusplus/draft#6049 .

**Comment by jensmaurer** [comment]
I've amended CWG2687 with the additional concerns.

**Comment by jensmaurer** [comment]
Looks like I'm a bit confused. I'll update the narrative, and we might need to see this again in CWG.

**Comment by LegionMammal978** [comment]
> Perhaps I am missing something here, but isn't that exactly what the proposed fix does? Per the (unchanged) rest of the paragraph, `(&x)(...)` is still equivalent to `(x)(...)`, so the object argument is still implicit (which is surprising). Making this work as you described would require more drastic changes to [over].

Hmm, true. I guess the problem here is that explicit object member functions would have to have two modes of overload resolution, one receiving the object argument from a class member access or from `(*this)`, and the other requiring the object argument to be received explicitly. Most directly, it seems like this could be patched a change to [[over.call.func]/3](https://wg21.link/over.call.func#3), to place explicit object member functions into case (1) instead of case (2), only when called by address. But then I have no clue how it would proceed when the lookup finds both explicit object member functions and static member functions, the latter of which has to receive a contrived object argument.

Perhaps this change would be a bad idea after all. (I wonder now if this trouble was forseen when P0847 was written.)

**Comment by jensmaurer** [comment]
I think the rationale for CWG2687 is that `(&A::f)(something)` cannot possibly be valid if `f` is an implicit object member function (because we're trying to directly call a pointer-to-member), but it is (or can be) valid if `f` is an explicit object member function, because `&A::f` is a plain pointer-to-function. I've updated CWG2687 accordingly.

CWG2692 addresses the concern about "lookup finds both explicit object member functions and static member functions".
